crates/red_knot_python_semantic/src/types.rs

```diff
@@ -82,10 +89,24 @@ pub(crate) fn definition_ty<'db>(db: &'db dyn Db, definition: Definition<'db>) -
 /// provide an `unbound_ty`.
 pub(crate) fn definitions_ty<'db>(
     db: &'db dyn Db,
-    definitions: &[Definition<'db>],
+    definitions_with_constraints: DefinitionWithConstraintsIterator<'_, 'db>,
     unbound_ty: Option<Type<'db>>,
 ) -> Type<'db> {
-    let def_types = definitions.iter().map(|def| definition_ty(db, *def));
+    let def_types = definitions_with_constraints.map(
+        |DefinitionWithConstraints {
+             definition,
+             constraints,
+         }| {
+            let mut builder = IntersectionBuilder::new(db);
+            builder = builder.add_positive(definition_ty(db, definition));
+            for constraint_ty in
+                constraints.filter_map(|test| narrowing_constraint(db, test, definition))
+            {
+                builder = builder.add_positive(constraint_ty);
+            }
+            builder.build()
+        },
+    );
     let mut all_types = unbound_ty.into_iter().chain(def_types);
```

Comment:
Not sure if this method shows up in the profiles but the `IntersectionBuilder` now always allocates an `FxHashSet` even for definitions with no constraints.

Reply:
I'll check the profile. Could easily special-case no-constraints here to avoid this.